repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -82,21 +88,57 @@ class Session(interpreter: Interpreter, stateChangedCallback: SessionState => Un
   def execute(code: String): Int = {
     val statementId = newStatementId.getAndIncrement()
     _statements(statementId) = new Statement(statementId, StatementState.Waiting, null)
+
     Future {
-      _statements(statementId) = new Statement(statementId, StatementState.Running, null)
+      def stmtState = _statements(statementId).state
+
+      if (stmtState != StatementState.Cancelled) {
+        lock.synchronized {
```

Comment:
We can get rid of this global lock if we make `Statement` mutable and lock the `Statement` being executed/cancelled.